Added new command 'spo list add' solving issue #204

rfjschouten opened the pull request:
Added new command 'spo list add' solving issue #204

waldekmastykarz commented:
Thanks @rfjschouten! I'll have a look asap!

waldekmastykarz commented:
@rfjschouten when trying to run `o365 spo list add -t Announcements --baseTemplate 107 -u https://contoso.sharepoint.com` I'm getting the following error:

```
TypeError: args.options.baseTemplate.trim is not a function
```

I suspect it's because the baseTemplate value is treated as a number rather than a string. Could you have a look at it and update the PR?

rfjschouten commented:
I will have a look at it Waldek. I think I'll need to check if the given value is typeof string. I will have to do that too for other enum options I defined. I will update the PR.

rfjschouten commented:
@waldekmastykarz Updated PR with checks on typeof string for enum options in command and added tests for it.

waldekmastykarz commented:
Thanks @rfjschouten. I'll have a look asap!

waldekmastykarz commented:
@rfjschouten if I run `o365 spo list add -t Announcements --baseTemplate 107 -u https://contoso.sharepoint.com` now, I'm getting the following error:

```
BaseTemplate option 107 is not recognized as valid choice. Please note it is case sensitive
```

Could it be that examples in command help need updating to use enums instead of numbers?

rfjschouten commented:
@waldekmastykarz Sorry about that. I updated the help command with examples using enums instead of numbers. Also added extra links to enum information in list-add.md file.

waldekmastykarz commented:
Thanks @rfjschouten. I'll recheck asap

waldekmastykarz commented:
Hey @rfjschouten, I've found one more issue. When specifying command options, you define a number of them as boolean:

```
interface Options extends GlobalOptions {
  title: string;
  baseTemplate: string;
  webUrl: string;
  description?: string;
  templateFeatureId?: string;
  schemaXml?: string;
  allowDeletion?: boolean;
  allowEveryoneViewItems?: boolean;
  allowMultiResponses?: boolean;
  contentTypesEnabled?: boolean;
  crawlNonDefaultViews?: boolean;
  ...
}

Despite the TypeScript definition, all these values are passed internally as string or number (try using `123` as value for `allowDeletion` and look at the value passed in the request when using the `--debug` option). If true|false are the only allowed values, we should validate them before executing the command the same way we do for GUIDs or URLs.

Also, thinking about the telemetry: an optional boolean value has three states: true, false and not set. With the following we're tracking only two of them, true and false:

```
telemetryProps.allowDeletion = args.options.allowDeletion || false;
```

How about we changed this to

```
if (args.options.allowDeletion) {
  telemetryProps.allowDeletion = args.options.allowDeletion === 'true';
}
```

That way, if the property is not set, we won't include it in the telemetry.

rfjschouten commented:
@waldekmastykarz good points! 
Maybe this is a little bit out of range for this issue, but more an overall addition for the CLI: how are boolean values validated and passed to commands. 
I see different implementations in the CLI. For instance 'spo cdn set' the option --enabled has an autocomplete. autocomplete: ['true', 'false'] and then in validate(): 

const enabled: string | undefined = args.options.enabled ? args.options.enabled.toLowerCase() : undefined;
      if (enabled !== 'true' &&
        enabled !== 'false') {
        return `${args.options.enabled} is not a valid boolean value. Allowed values are true|false`;
      }

If I pass in the option --enabled 103 to this command, it will fail on .toLowerCase()

For an end user using the Office 365 CLI it is confusing on how to pass boolean values to commands, because of different implementations. 

So I think it is a good thing to determine a default way for passing boolean values. In my command you can also only pass in the option like spo list add --allowDeletion without passing true or false. Only passing the option tells the command to set it to true. I like that.

So what is a good option to pass and validate boolean values to the commands?

rfjschouten commented:
On the telemetry part, I think you mean:
```
    if (args.options.allowDeletion) {
      telemetryProps.allowDeletion = args.options.allowDeletion === true;
    }
```

instead of:

```
    if (args.options.allowDeletion) {
      telemetryProps.allowDeletion = args.options.allowDeletion === 'true';
    }
```

allowDeletion is a boolean. Good addition on the extra check to only add it to the telemetry if it is set. 👍🏻

waldekmastykarz commented:
@rfjschouten you're right that we might have missed a few places where we deal with booleans. My current understanding is as follows:

- if you define a switch option (`--verbose`) the option value is undefined or boolean true
- if you have an option that should allow boolean, like `--enabled` then the value provided by the user is passed as either a string or a number, as you noticed, but never a boolean. Using types you can force the value to be always a string, but never a boolean, so to perform boolean comparisons, the value has to be cast first.
- alternatively, we could make use of the capability to implement boolean values, using switches. In that case a `false` value can be provided using the `no` prefix, eg. `--enabled` > true, `--no-enabled` > false. This however seems a not intuitive to me and I haven't seen much use of it. using `--enabled false` seems clearer to me, but means some extra work on our end.

rfjschouten commented:
@waldekmastykarz I changed as per requested. Changed all boolean options into string type, added autocomplete [true,false] to those options, wrote validations for it and tests. I added new Utils method isValidBoolean and tests for it.

waldekmastykarz commented:
Thanks! I'll have a look shortly

waldekmastykarz commented:
Merged manually. Thanks for the contribution @rfjschouten!

rfjschouten commented:
My pleasure @waldekmastykarz ! Up to the next one!

waldekmastykarz commented:
Awesome! Appreciate your help 💪
